### ISSUE TITLE
Bootstrapped Neural Networks RFs and the Mushroom bandit

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10805]
Bootstrapped Neural Networks, RFs, and the Mushroom bandit<br>
<br>
https://ift.tt/n36pOSJ<br>
<br>
Guilherme Duarte Marmerola<br>
<br>

<div><div><div><div><article><p>Thompson Sampling is a very simple yet effective method to addressing the exploration-exploitation dilemma in reinforcement/online learning. In this series of posts, I’ll introduce some applications of Thompson Sampling in simple examples, trying to show some cool visuals along the way. All the code can be found on my GitHub page <a href="https://github.com/gdmarmerola/interactive-intro-rl">here</a>.</p>
<p>In this post, we explore more advanced algorithms, starting with a simple trial to get uncertainties out of Random Forests and then moving to Bootstrapped Neural Networks. We use these methods to solve the “Mushroom” bandit, a contextual bandit game where the agent must decide, from a bucket of edible and poisonous mushrooms, which ones to eat.</p>
<h2>Solving the Mushroom Bandit</h2>
<p>In this notebook, we’ll run an experiment inspired by the paper <a href="https://arxiv.org/pdf/1505.05424.pdf">Weight Uncertainty in Neural Networks</a> (Blundell et. al, 2015), the Mushroom bandit. In this setting, we use the Mushroom dataset from UCI to simulate a bandit-like game, in order to test some Thompson Sampling-inspired algorithms that try to address the exploration-exploitation tradeoff. Particularly, we will test two interesting ideas:</p>
<ol>
<li><p><strong>Bootstrapped Neural Networks:</strong> based on Bootstrapped DQN, a proven method for <em>deep exploration</em> in reinforcement learning, presented by <a href="https://arxiv.org/abs/1602.04621">Osband et. al. (2016)</a>. Here, we do not have a full RL setting, so we call it a Bootstrapped Neural Network, rather than a Bootstrapped DQN.</p></li>
<li><p><strong>Sampling from a Random Forest:</strong> let us try to use the individual tree predictions of a Random Forest to build uncertainty estimates for our game.</p></li>
</ol>
<p>We will also test the greedy counterparts of these ideas, a regular neural network and a regular Random Forest.</p>
<h2>Test with toy data</h2>
<p>Let us first run our models with simple toy data to check uncertainty estimates.</p>
<h3>Generating the data</h3>
<p>We will use the formula given by the paper <a href="https://arxiv.org/pdf/1505.05424.pdf"><em>Weight Uncertainty in Neural Networks</em></a> to generate the data:</p>\[y = x + 0.3 \cdot{} sin(2π(x + \epsilon)) + 0.3 \cdot{} sin(4 \cdot{} \pi \cdot{}(x + \epsilon)) + \epsilon\]<p>where $\epsilon \sim \mathcal{N}(0, 0.02)$. We take a hundred $x$ samples from an uniform distribution $\mathcal{U}(0, 0.5)$. This results in the following plot:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-1.png"></p>
<h3>Random Forest</h3>
<p>Let us start with the simpler model, a Random Forest. What we will try to do is randomize the forest’s output, by sampling sets of decision trees and averaging their predictions. The following code accomplishes that, drawing 20 samples of our approximate posterior distribution.</p>
<pre><code># instance of RF
et = ExtraTreesRegressor(n_estimators=100, min_samples_leaf=2)

# let us fit it to the data
et.fit(x.reshape(-1,1),y)

# let us generate data to check the fit
x_grid = np.linspace(-0.5, 1.0, 200).reshape(-1,1)

# number of samples from the forest
N_SAMPLES = 20

# let us get the predictions of four trees at a time
tree_samples = [np.random.choice(et.estimators_, 4) for i in range(N_SAMPLES)]

# let us predict with these guys
y_grid = np.array([np.array([e.predict(x_grid) for e in np.random.choice(et.estimators_, 2)]).mean(axis=0) for tree_sample in tree_samples])
</code></pre>
<p>The sampling procedure results in the following plot:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-2.png"></p>
<p>The uncertainty estimates concentrate where there is more data, which is a good sign. However, uncertainties do not increase as we get distant from the samples. This may be due to the rule-based learning procedure of RF’s, which prevents them from extrapolating.</p>
<h3>Bootstrapped Neural Network</h3>
<p>Let us check the uncertainty estimates produced by our Bootstrapped NN. As introduced by <a href="https://arxiv.org/abs/1602.04621">Osband et. al. (2016)</a>, we will train a network with a shared layer and many bootstrapped heads, which will diverge a little bit on their predictions. This divergence will drive efficient exploration and will give us reasonable uncertainty estimates.</p>
<p>I made a quick and dirty implementation of this model on Keras, but it’s far from ideal. Here, I generate the bootstrap samples at model compile time, and return $N$ bootstrapped samples of the training dataset, consuming a lot of memory. The model trains on 10 different inputs and outputs.</p>
<div><div><pre><code><span># defining the model #
</span>
<span># shared network
</span><span>shared_net</span> <span>=</span> <span>Sequential</span><span>([</span><span>Dense</span><span>(</span><span>1</span><span>,</span> <span>activation</span><span>=</span><span>'linear'</span><span>,</span> <span>input_dim</span><span>=</span><span>1</span><span>),</span> 
                         <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>),</span> 
                         <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>),</span>
                         <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>)])</span>

<span># bootstrap rate and number of heads
</span><span>BOOTSTRAP_RATE</span> <span>=</span> <span>1.00</span>
<span>N_HEADS</span> <span>=</span> <span>10</span>

<span># initializing lists </span><span>inputs</span> <span>=</span> <span>[]</span>
<span>heads</span> <span>=</span> <span>[]</span>
<span>x_btrap</span> <span>=</span> <span>[]</span>
<span>y_btrap</span> <span>=</span> <span>[]</span>

<span># loop for creating heads and sampled data points
</span><span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>N_HEADS</span><span>):</span>
    
    <span># deciding sample of data points
</span>    <span>sample_idx</span> <span>=</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>choice</span><span>(</span><span>range</span><span>(</span><span>x</span><span>.</span><span>shape</span><span>[</span><span>0</span><span>]),</span> <span>int</span><span>(</span><span>x</span><span>.</span><span>shape</span><span>[</span><span>0</span><span>]</span><span>*</span><span>BOOTSTRAP_RATE</span><span>),</span> <span>True</span><span>)</span>
    
    <span># defining the heads in the graph
</span>    <span>input_temp</span> <span>=</span> <span>Input</span><span>(</span><span>shape</span><span>=</span><span>(</span><span>1</span><span>,))</span>
    <span>shared_temp</span> <span>=</span> <span>shared_net</span><span>(</span><span>input_temp</span><span>)</span>
    <span>head_temp</span> <span>=</span> <span>Dense</span><span>(</span><span>16</span><span>,</span> <span>activation</span><span>=</span><span>'relu'</span><span>)(</span><span>shared_temp</span><span>)</span>
    <span>heads</span><span>.</span><span>append</span><span>(</span><span>Dense</span><span>(</span><span>1</span><span>,</span> <span>activation</span><span>=</span><span>'linear'</span><span>)(</span><span>head_temp</span><span>))</span>
    <span>inputs</span><span>.</span><span>append</span><span>(</span><span>input_temp</span><span>)</span>
    
    <span># defining data
</span>    <span>x_btrap</span><span>.</span><span>append</span><span>(</span><span>x</span><span>[</span><span>sample_idx</span><span>])</span>
    <span>y_btrap</span><span>.</span><span>append</span><span>(</span><span>y</span><span>[</span><span>sample_idx</span><span>])</span>
    
<span># defining the model and plotting it
</span><span>model</span> <span>=</span> <span>Model</span><span>(</span><span>inputs</span><span>=</span><span>inputs</span><span>,</span> <span>outputs</span><span>=</span><span>heads</span><span>)</span>
<span>plot_model</span><span>(</span><span>model</span><span>)</span>

<span># compiling
</span><span>model</span><span>.</span><span>compile</span><span>(</span><span>loss</span><span>=</span><span>'mean_squared_error'</span><span>,</span> <span>optimizer</span><span>=</span><span>'adam'</span><span>,</span> <span>metrics</span><span>=</span><span>[</span><span>'mean_squared_error'</span><span>])</span>

<span># fitting the model
</span><span>model</span><span>.</span><span>fit</span><span>(</span><span>x_btrap</span><span>,</span> <span>y_btrap</span><span>,</span> <span>epochs</span><span>=</span><span>500</span><span>,</span> <span>batch_size</span><span>=</span><span>64</span><span>,</span> <span>verbose</span><span>=</span><span>0</span><span>);</span>
</code></pre></div></div>
<p>Even with my dirty code, the results are beautiful. We can see a reasonable uncertainty estimate given by the individual predictions of the bootstrapped heads:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-3.png"></p>
<p>We are now ready to play the Mushroom bandit game!</p>
<h3>The Mushroom bandit</h3>
<p>The game is set up as follows:</p>
<ol>
<li>The agent is initialized with 50 random mushrooms and if they’re edible or poisonous</li>
<li>At each round, the agent is presented with all the mushrooms which were not already eaten.</li>
<li>The agent ranks the mushrooms, with the most edible mushroom at the top.</li>
<li>The agent eats the top $k$ mushrooms and receives feedback. The agent reward is the actual observation.</li>
</ol>
<p>Here’s my code to implement the game:</p>
<div><div><pre><code><span># coding a class for this
</span><span>class</span> <span>MushroomBandit</span><span>:</span>
    
    <span># initializing
</span>    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>,</span> <span>X</span><span>,</span> <span>y</span><span>):</span>
        
        <span># storing the initialization
</span>        <span>self</span><span>.</span><span>X</span> <span>=</span> <span>X</span><span>;</span> <span>self</span><span>.</span><span>y</span> <span>=</span> <span>y</span>
        
        <span># let us keep eaten and not eaten mushrooms as well
</span>        <span>self</span><span>.</span><span>X_not_eaten</span> <span>=</span> <span>X</span><span>.</span><span>copy</span><span>()</span>
        <span>self</span><span>.</span><span>y_not_eaten</span> <span>=</span> <span>y</span><span>.</span><span>copy</span><span>()</span>
        <span>self</span><span>.</span><span>X_eaten</span> <span>=</span> <span>pd</span><span>.</span><span>DataFrame</span><span>()</span>
        <span>self</span><span>.</span><span>y_eaten</span> <span>=</span> <span>pd</span><span>.</span><span>Series</span><span>()</span>
        
    <span># function to show not eaten mushrooms
</span>    <span>def</span> <span>get_not_eaten</span><span>(</span><span>self</span><span>):</span>
        <span>return</span> <span>self</span><span>.</span><span>X_not_eaten</span>
    
    <span># function to show eaten mushrrooms
</span>    <span>def</span> <span>get_eaten</span><span>(</span><span>self</span><span>):</span>
        <span>return</span> <span>self</span><span>.</span><span>X_eaten</span><span>,</span> <span>self</span><span>.</span><span>y_eaten</span>
    
    <span># function to get which mushrooms will be eaten
</span>    <span>def</span> <span>eat_mushrooms</span><span>(</span><span>self</span><span>,</span> <span>mushroom_idx</span><span>):</span>
        
        <span># get feedback
</span>        <span>feedback</span> <span>=</span> <span>self</span><span>.</span><span>y_not_eaten</span><span>.</span><span>loc</span><span>[</span><span>mushroom_idx</span><span>].</span><span>copy</span><span>()</span>
        
        <span># remove eaten mushrooms from not eaten dataset
</span>        <span>self</span><span>.</span><span>X_not_eaten</span> <span>=</span> <span>self</span><span>.</span><span>X_not_eaten</span><span>.</span><span>drop</span><span>(</span><span>mushroom_idx</span><span>)</span>
        <span>self</span><span>.</span><span>y_not_eaten</span> <span>=</span> <span>self</span><span>.</span><span>y_not_eaten</span><span>.</span><span>drop</span><span>(</span><span>mushroom_idx</span><span>)</span>
        
        <span># add eaten mushrooms from eaten dataset
</span>        <span>self</span><span>.</span><span>X_eaten</span> <span>=</span> <span>pd</span><span>.</span><span>concat</span><span>([</span><span>self</span><span>.</span><span>X_eaten</span><span>,</span> <span>self</span><span>.</span><span>X</span><span>.</span><span>loc</span><span>[</span><span>mushroom_idx</span><span>]])</span>
        <span>self</span><span>.</span><span>y_eaten</span> <span>=</span> <span>pd</span><span>.</span><span>concat</span><span>([</span><span>self</span><span>.</span><span>y_eaten</span><span>,</span> <span>self</span><span>.</span><span>y</span><span>.</span><span>loc</span><span>[</span><span>mushroom_idx</span><span>]])</span>
        
        <span># return feedback
</span>        <span>return</span> <span>feedback</span>
</code></pre></div></div>
<p>We first play the game with a random policy. The code for running an episode of the game is very simple:</p>
<div><div><pre><code><span># number of mushrooms eaten per round
</span><span>K</span> <span>=</span> <span>100</span>

<span># instance of the game
</span><span>mb</span> <span>=</span> <span>MushroomBandit</span><span>(</span><span>X</span><span>,</span> <span>y</span><span>)</span>

<span># we initialize the agent by eating the first 50 mushrooms of the base
</span><span>round_feedback</span> <span>=</span> <span>mb</span><span>.</span><span>eat_mushrooms</span><span>(</span><span>mb</span><span>.</span><span>get_not_eaten</span><span>().</span><span>head</span><span>(</span><span>50</span><span>).</span><span>index</span><span>)</span>

<span># and we get the number of rounds
</span><span>n_rounds</span> <span>=</span> <span>int</span><span>(</span><span>mb</span><span>.</span><span>get_not_eaten</span><span>().</span><span>shape</span><span>[</span><span>0</span><span>]</span><span>/</span><span>K</span><span>)</span> <span>+</span> <span>1</span>

<span># list and to accumulate rewards
</span><span>cum_rewards</span> <span>=</span> <span>0</span>
<span>cum_rewards_list_random</span> <span>=</span> <span>[]</span>

<span># then, we play until no mushrooms are left
</span><span>for</span> <span>round_id</span> <span>in</span> <span>tqdm</span><span>(</span><span>range</span><span>(</span><span>n_rounds</span><span>)):</span>
    
    <span># choosing random mushrooms
</span>    <span>to_eat</span> <span>=</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>choice</span><span>(</span><span>mb</span><span>.</span><span>get_not_eaten</span><span>().</span><span>index</span><span>,</span> <span>K</span><span>)</span>
    
    <span># eating
</span>    <span>feedback</span> <span>=</span> <span>mb</span><span>.</span><span>eat_mushrooms</span><span>(</span><span>to_eat</span><span>)</span>
    
    <span># saving cumulative rewards
</span>    <span>cum_rewards</span> <span>+=</span> <span>feedback</span><span>.</span><span>sum</span><span>()</span>
    <span>cum_rewards_list_random</span><span>.</span><span>append</span><span>(</span><span>cum_rewards</span><span>)</span>
</code></pre></div></div>
<p>At the end of game we inspect the number of edible mushrooms eaten during the course of the episode. The random policy fares as we would expect: a straight line, as it eats edible mushrooms at the same rate as poisonous ones (considering class imbalance).</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-4.png"></p>
<p>We then proceed to replacing the random policy with a regular neural network and our bootstrapped neural network. The regular net is the same as a greedy method, as it only uses the expectation learned by the net to make decisions. The bootstrapped net emulates Thompson Sampling, as each head acts like a sample of a posterior distribution. Thus, we expect that the BNN explores efficiently and makes better decisions than the regular net. The plot below shows the results achieved:</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-5.png"></p>
<p>As we would expect, the boostrapped net finds all the edible mushrooms faster than the regular net, as it explores more efficiently. This technique provides a good bayesian approximation to neural networks, and can be used in a number of other settings. There’s great discussion about bayesian deep learning right now, particularly concerning the <a href="https://pdfs.semanticscholar.org/dde4/b95be20a160253a6cc9ecd75492a13d60c10.pdf">advantages of this approach compared to Monte Carlo dropout</a>. I look forward to seeing the evolution of the field in the next years.</p>
<p>Finally, we run a regular (greedy) RF against our sampled RF. Interestingly, both RF models beat neural networks, with the sampled RF overcoming the greedy RF only at the very end of game. This is a very curious result, which teases us to test the models in other datasets. If you want to do that, remember that the code for this experiment is freely available at my GitHub page.</p>
<p><img src="https://gdmarmerola.github.io//assets/img/ts_for_mushroom_bandit/mushroom-img-6.png"></p>
<h2>Conclusion</h2>
<p>Even that the mushroom bandit was easily solved by the models, we can check that randomizing decisions offer improvements over greedy decisions. On the final plot the randomized models outperformed the greedy models, especially on the hardest mushrooms (the last ones to be eaten). In this particular case, Random Forest compared favourably against neural networks, and are trained and sampled in a fraction of the time.</p>
<p>In the future, we should try out other methods like UCB and $\epsilon$-greedy, and check if they will be outperformed by Thompson Sampling, as recent research suggests.</p></article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Guilherme’s Blog <br>
<br>
February 5, 2024 at 04:53PM
